linux/ibus-keyman/tests/scripts/test-helper.inc.sh

```diff
@@ -0,0 +1,229 @@
+      FIRST=false
+    else
+      echo "    ," >> "$KMPFILE"
+    fi
+    cat <<-EOF >> "$KMPFILE"
+    {
+      "name": "$keyboard",
+      "id": "$keyboard",
+      "version": "0.0",
+      "languages": [
+        {
+          "name": "Undetermined",
+          "id": "und"
+        }
+      ]
+    }
+EOF
+  done
+  echo "  ]" >> "$KMPFILE"
+  echo "}" >> "$KMPFILE"
+  popd > /dev/null || exit
+}
+
+function link_test_keyboards() {
+  KMX_TEST_DIR=$1
+  TESTDIR=$2
+  TESTBASEDIR=$3
+
+  if [[ $(find "${KMX_TEST_DIR}/" -name k_\*.kmx 2>/dev/null | wc -l) -gt 0 ]]; then
+    mkdir -p "$(realpath --canonicalize-missing "$TESTBASEDIR")"
+    rm -f "$TESTDIR"
+    ln -sf "$(realpath "${KMX_TEST_DIR}")" "$TESTDIR"
+  else
+    echo "Can't find test kmx files in ${KMX_TEST_DIR}"
+    exit 3
+  fi
+}
+
+function setup() {
+  local DISPLAY_SERVER ENV_FILE PID_FILE TOP_SRCDIR TOP_BINDIR TESTBASEDIR TESTDIR
+  DISPLAY_SERVER=$1
+  ENV_FILE=$2
+  PID_FILE=$3
+
+  TOP_SRCDIR=${G_TEST_SRCDIR:-$(realpath "$(dirname "$0")/..")}/..
+  TOP_BINDIR=${G_TEST_BUILDDIR:-$(realpath "$(dirname "$0/..")")}/..
+  TESTBASEDIR=${XDG_DATA_HOME:-$HOME/.local/share}/keyman
+  TESTDIR=${TESTBASEDIR}/test_kmx
+
+  echo > "$ENV_FILE"
+
+  if [ -f "$PID_FILE" ]; then
+    # kill previous instances
+    "$(dirname "$0")"/teardown-tests.sh "$PID_FILE"
+  fi
+
+  echo > "$PID_FILE"
+  TEMP_DATA_DIR=$(mktemp --directory)
+  echo "rm -rf ${TEMP_DATA_DIR} || true" >> "$PID_FILE"
+
+  COMMON_ARCH_DIR=
+  [ -d "${TOP_SRCDIR}"/../../core/build/arch ] && COMMON_ARCH_DIR=${TOP_SRCDIR}/../../core/build/arch
+
+  if [ -d "${COMMON_ARCH_DIR}"/release ]; then
+    COMMON_ARCH_DIR=${COMMON_ARCH_DIR}/release
+  elif [ -d "${COMMON_ARCH_DIR}"/debug ]; then
+    COMMON_ARCH_DIR=${COMMON_ARCH_DIR}/debug
+  else
+    echo "Can't find neither ${COMMON_ARCH_DIR}/release nor ${COMMON_ARCH_DIR}/debug"
+    exit 2
+  fi
+
+  link_test_keyboards "${TOP_SRCDIR}/../../common/test/keyboards/baseline" "$TESTDIR" "$TESTBASEDIR"
+
+  generate_kmpjson "$TESTDIR"
+
+  # Start test dbus server
+  "${TOP_BINDIR}/tests/km-dbus-test-server" &> /dev/null &
+  sleep 1
+  source /tmp/km-test-server.env
+  cat /tmp/km-test-server.env >> "$ENV_FILE"
+  cat /tmp/km-test-server.env >> "$PID_FILE"
+  echo "${TOP_BINDIR}/tests/stop-test-server" >> "$PID_FILE"
+
+  if [ "$DISPLAY_SERVER" == "wayland" ]; then
+    if ! can_run_wayland; then
+      # support for --headless got added in mutter 40.x
+      echo "ERROR: mutter doesn't support running headless. Can't run Wayland tests."
+      exit 7
+    fi
+    echo "Running on Wayland:"
+    echo "Starting mutter..."
+    TMPFILE=$(mktemp)
+    # mutter-Message: 18:56:15.422: Using Wayland display name 'wayland-1'
+    mutter --wayland --headless --no-x11 --virtual-monitor 1024x768 &> "$TMPFILE" &
+    echo "kill -9 $! || true" >> "$PID_FILE"
+    sleep 1s
+    export WAYLAND_DISPLAY
+    WAYLAND_DISPLAY=$(grep "Using Wayland display" "$TMPFILE" | cut -d"'" -f2)
```

Comment:
I notice this is how the previous code was. But why wouldn't the two lines be combined?
```suggestion
    export WAYLAND_DISPLAY=$(grep "Using Wayland display" "$TMPFILE" | cut -d"'" -f2)
```

Reply:
Combining it masks failures. See https://www.shellcheck.net/wiki/SC2155.